src/initial_conditions.jl

```diff
@@ -346,6 +373,25 @@ function enforce_z_boundary_condition!(f, bc::String, adv::T, vpa, r, compositio
     end
 end
 
+"""
+enforce the z boundary condition on the evolved velocity space moments of f
+"""
+function enforce_z_boundary_condition_moments!(density, moments, bc::String)
+    # TODO: parallelise
+    @serial_region begin
+        # enforce z boundary condition on density if it is evolved separately from f
+    	if moments.evolve_density
+            # TODO: extend to 'periodic' BC case, as this requires further code modifications to be consistent
+            # with finite difference derivatives (should be fine for Chebyshev)
+            if bc == "wall"
+                @loop_s_r is ir begin
+                    density[1,ir,is] = 0.5*(density[1,ir,is] + density[end,ir,is])
+                    density[end,ir,is] = density[1,ir,is]
```

Comment:
This is the best BC I could think of, but am open to improvements.  Here is my reasoning:  we have zero incoming BC for ions and BC for neutrals that any ions/neutrals leaving the domain re-enter as neutrals.  While this is a fine BC for the pdf, it does not provide a definite BC for the density.  I used the fact that the drift kinetic/quasineutrality system is such that if f(z,vpa,t) and phi(z,t) are solutions, then so too are f(-z,-vpa,t) and phi(-z,t).  In a statistical sense, this means that the solution for f should be even in the (z,vpa) plane.  This would mean that the density at one wall is equal to the density at the other wall, so I chose to enforce periodicity.  For upar, we would have that parallel flow is equal and opposite at the two walls.

Reply:
That does make sense.

Could we get away with not putting any boundary condition on the density? That is what we try to do with STORM (by extrapolating into boundary cells rather than imposing a specific boundary condition), with the justification that for the pair of density/parallel-velocity the velocity has 2 boundary conditions (one at each boundary), while (neglecting parallel viscosity) there are 2 parallel derivatives between the two equations, so we 'should' (?) need 2 boundary conditions in total.

Or in other words, for moment_kinetics, the boundary condition on the distribution function fixes the flux of density into the boundary, and maybe that is enough of a boundary condition without doing anything to the density explicitly?

Reply:
To be honest, I'm not sure which is better.  Because we are evolving g=f / n which satisfies int dv g = 1, enforcing BCs on g does not appear to fix the particle flux at the boundary but rather the flow (density has been removed).  Once we move on to using the peculiar velocity (next up for me), BCs on g won't even be fixing that much.  That said, we've been running without explicit BCs on the fluid quantities thus far without problems, and it's not clear how to generalise the BC I've proposed for cases where there is some asymmetry  in the walls at the two domain boundaries...

Reply:
Sorry, I tend to interchange 'flux' and 'flow' in my head, and really shouldn't! In STORM the actual boundary condition is on flow (we use ion parallel velocity equal to sound speed), which seems to work, although I've never known whether we only get away with things by having a fairly large amount of numerical dissipation...

When evolving peculiar velocity, it does seem likely we'd need a bc for the parallel velocity (or flux?) and I don't have a clue what that should be...

Reply:
In Felix's "REPORT 2047357-TN-05-02 M1.3", titled "1D drift kinetic models with wall boundary conditions", the suggestion was:
> These equations for ions and neutrals have to be solved with the boundary conditions in equations (3.7) and (3.8). For n_s, upar_s, vt_s and F_s known at time t, we can construct f_s at z = 0 and z = L, and we can apply boundary conditions (3.7) and (3.8). We can then use the resulting f_s to obtain n_s, upar_s, vt_s and F_s, and to calculate p_s and q_s, closing the system of equations.
> 
> We note that this is the only place where the full distribution function f_s is needed. Depending on the numerical method chosen to solve these equations, reconstructing f_s could be expensive. This is one of the problems that needs to be addressed in the proxy apps.

Reply:
I will need to look at the actual implementation, but it seems to me like this will be equivalent to not imposing any separate BCs on density and parallel flow.  This is because, e.g., imposing f_i = 0 for vpa > 0 at z=left boundary and similar for right boundary is equivalent to imposing g_i = 0 for vpa > 0 at left boundary, etc.  I don't see how this gives any additional constraint on the density but need to look more carefully.  Can I propose for now that we merge this branch with master, and before I merge the next branch I create to evolve parallel flow with master we revisit?

Reply:
Sounds good to me!